core-rust/state-manager/src/limits.rs

```diff
@@ -0,0 +1,179 @@
+            total_transactions_size: 0,
+            total_execution_cost_units_consumed: 0,
+            total_substate_read_size: 0,
+            total_substate_read_count: 0,
+            total_substate_write_size: 0,
+            total_substate_write_count: 0,
+        }
+    }
+
+    pub fn check_pre_execution(
+        &self,
+        limits_config: &VertexLimitsConfig,
+        transaction_size: usize,
+    ) -> Result<(), VertexLimitsExceeded> {
+        if self.total_transactions_count + 1 > limits_config.max_total_transactions_count {
+            return Err(VertexLimitsExceeded::TransactionsCount);
+        }
+
+        if self.total_substate_read_size + transaction_size
+            > limits_config.max_total_transactions_size
+        {
+            return Err(VertexLimitsExceeded::TransactionsSize);
+        }
+
+        Ok(())
+    }
+
+    pub fn check_post_execution(
+        &self,
+        limits_config: &VertexLimitsConfig,
+        execution_metrics: &ExecutionMetrics,
+    ) -> Result<(), VertexLimitsExceeded> {
+        if self.total_execution_cost_units_consumed
+            + execution_metrics.execution_cost_units_consumed
+            > limits_config.max_total_execution_cost_units_consumed
+        {
+            return Err(VertexLimitsExceeded::ExecutionCostUnitsConsumed);
+        }
+
+        if self.total_substate_read_size + execution_metrics.substate_read_size
+            > limits_config.max_total_substate_read_size
+        {
+            return Err(VertexLimitsExceeded::SubstateReadSize);
+        }
+
+        if self.total_substate_read_count + execution_metrics.substate_read_count
+            > limits_config.max_total_substate_read_count
+        {
+            return Err(VertexLimitsExceeded::SubstateReadCount);
+        }
+
+        if self.total_substate_write_size + execution_metrics.substate_write_size
+            > limits_config.max_total_substate_write_size
+        {
+            return Err(VertexLimitsExceeded::SubstateWriteSize);
+        }
+
+        if self.total_substate_write_count + execution_metrics.substate_write_count
+            > limits_config.max_total_substate_write_count
+        {
+            return Err(VertexLimitsExceeded::SubstateWriteCount);
+        }
+
+        Ok(())
+    }
+
+    pub fn advance(&mut self, transaction_size: usize, execution_metrics: &ExecutionMetrics) {
+        self.total_transactions_count += 1;
+        self.total_transactions_size += transaction_size;
+        self.total_execution_cost_units_consumed += execution_metrics.execution_cost_units_consumed;
+        self.total_substate_read_count += execution_metrics.substate_read_count;
+        self.total_substate_read_size += execution_metrics.substate_read_size;
+        self.total_substate_write_count += execution_metrics.substate_write_count;
+        self.total_substate_write_size += execution_metrics.substate_write_size;
+    }
+
+    pub fn is_full(&self, limits_config: &VertexLimitsConfig) -> bool {
```

Comment:
The `limits_config` seems to be a param to almost all methods, so (e.g. in java) it would be no-brainer to make it an instance field. Is this difficult due to borrow-checker here?

Reply:
It was mere personal preference (pushing data ownership upwards if only read and let parent delegate strictly when needed). I'll capture it by ref.